homeassistant/components/tibber/sensor.py

```diff
@@ -410,86 +406,87 @@ def __init__(
         self._attr_state = state
         self.async_write_ha_state()
 
 
-class TibberRtDataHandler:
+class TibberRtDataCoordinator(update_coordinator.DataUpdateCoordinator):
     """Handle Tibber realtime data."""
 
     def __init__(self, async_add_entities, tibber_home, hass):
         """Initialize the data handler."""
         self._async_add_entities = async_add_entities
         self._tibber_home = tibber_home
         self.hass = hass
-        self._entities = {}
+        self._all_sensors = list(RT_SENSORS)
+        super().__init__(
+            hass,
+            _LOGGER,
+            name=tibber_home.info["viewer"]["home"]["address"].get(
+                "address1", "Tibber"
+            ),
+        )
 
-    async def async_callback(self, payload):
-        """Handle received data."""
-        errors = payload.get("errors")
-        if errors:
-            _LOGGER.error(errors[0])
-            return
-        data = payload.get("data")
-        if data is None:
-            return
-        live_measurement = data.get("liveMeasurement")
-        if live_measurement is None:
+        self._async_remove_device_updates_handler = self.async_add_listener(
+            self._add_sensors
+        )
+        hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, self._handle_ha_stop)
+
+    @callback
+    def _handle_ha_stop(self, _event) -> None:
+        """Handle Home Assistant stopping."""
+        self._async_remove_device_updates_handler()
+
+    @callback
+    def _add_sensors(self):
+        """Add sensor."""
+        if not (live_measurement := self.get_live_measurement()):
             return
 
-        timestamp = dt_util.parse_datetime(live_measurement.pop("timestamp"))
         new_entities = []
-        for sensor_type, state in live_measurement.items():
-            if state is None or sensor_type not in RT_SENSOR_MAP:
+        for sensor_description in self._all_sensors:
+            state = live_measurement.get(sensor_description.key)
+            if state is None:
                 continue
-            if sensor_type == "powerFactor":
-                state *= 100.0
-            if sensor_type in self._entities:
-                async_dispatcher_send(
-                    self.hass,
-                    SIGNAL_UPDATE_ENTITY.format(self._entities[sensor_type]),
-                    state,
-                    timestamp,
-                )
-            else:
-                entity = TibberSensorRT(
-                    self._tibber_home,
-                    RT_SENSOR_MAP[sensor_type],
-                    state,
-                )
-                new_entities.append(entity)
-                self._entities[sensor_type] = entity.unique_id
+            entity = TibberSensorRT(
+                self._tibber_home,
+                sensor_description,
+                state,
+                self,
+            )
+            new_entities.append(entity)
+            self._all_sensors.remove(sensor_description)
```

Comment:
We're modifying the iterable inside the loop.

Reply:
https://github.com/home-assistant/core/pull/53619/files#diff-2a4673e367cb32db74b949cd9c36f5ad31766f0147dc2673910879daa64b85d4R475

Reply:
Maybe instead have a separate set of added description keys and iterate all the descriptions and have a check to `continue` if the description iterable key matches an already added key. That way we don't need to make a copy every time we add sensors.